lib/rust/ensogl/core/src/display/object/class.rs

```diff
@@ -619,32 +694,168 @@ pub struct Id(usize);
+pub struct Instance<Host: 'static = Scene> {
     rc: Rc<Model<Host>>,
 }
 
-impl<Host> Deref for Instance<Host> {
-    type Target = Rc<Model<Host>>;
-    fn deref(&self) -> &Self::Target {
-        &self.rc
-    }
-}
 
 impl<Host> Instance<Host> {
-    /// Constructor.
-    pub fn new() -> Self {
-        default()
-    }
-
     /// Create a new weak pointer to this display object instance.
     pub fn downgrade(&self) -> WeakInstance<Host> {
         let weak = Rc::downgrade(&self.rc);
         WeakInstance { weak }
     }
 }
 
+impl<Host: 'static> Instance<Host> {
+    /// Constructor.
+    pub fn new() -> Self
+    where Host: 'static {
+        Self { rc: Rc::new(Model::new()) }.init()
+    }
+
+    fn init(self) -> Self
+    where Host: 'static {
+        let network = &self.network;
+        let this = &self;
+        frp::extend! { network
+            eval self.event_source ((event) this.emit_event_impl(event));
+        }
+        self
+    }
+
+    fn new_event<T>(&self, payload: T) -> event::SomeEvent
+    where
+        Host: 'static,
+        T: 'static, {
+        event::SomeEvent::new(Some(self), payload)
+    }
+
+    fn _emit_event<T>(&self, payload: T)
+    where
+        Host: 'static,
+        T: 'static, {
+        self.event_source.emit(event::SomeEvent::new(Some(self), payload));
+    }
+
+    fn emit_event_impl(&self, event: &event::SomeEvent) {
+        let rev_parent_chain = self.rev_parent_chain();
+        if event.captures.get() {
+            for object in &rev_parent_chain {
+                object.capturing_event_fan.emit(&event.data);
+                if event.state() == event::State::Cancelled {
+                    break;
+                }
+            }
+        }
+        if event.state() != event::State::Cancelled {
+            self.capturing_event_fan.emit(&event.data);
+        }
+        if event.state() != event::State::Cancelled {
+            self.bubbling_event_fan.emit(&event.data);
+        }
+        if event.bubbles.get() {
+            for object in rev_parent_chain.iter().rev() {
+                object.bubbling_event_fan.emit(&event.data);
+                if event.state() == event::State::Cancelled {
+                    break;
+                }
+            }
+        }
+    }
+
+    /// Get reversed parent chain of this display object (`[root, child_of root, ...,
+    /// parent_of_this_object]`). It will NOT contain this object.
+    pub fn rev_parent_chain(&self) -> Vec<Instance<Host>> {
+        let mut vec = default();
+        self.build_rev_parent_chain(&mut vec);
+        vec
+    }
+
+    fn build_rev_parent_chain(&self, vec: &mut Vec<Instance<Host>>) {
+        if let Some(parent) = self.parent() {
+            parent.build_rev_parent_chain(vec);
+            vec.push(parent);
+        }
+    }
+
+    fn focused_instance(&self) -> Option<Instance<Host>> {
+        if let Some(child) = &*self.focused_descendant.borrow() {
+            child.upgrade()
```

Comment:
I'm sorry, would you be so nice to rephrase your comment? I did not get its meaning.

Reply:
If there's no child because `focused_descendant` is `None`, the logic in the following `else` branch runs--but if there's no child because `focused_descendant` is an expired weak handle, we're returning `None` (the result of the `upgrade` call) without entering the `else` branch. It's not obvious to me whether this inconsistency is intentional.

Reply:
Oh, gotcha. Right, I already fixed it before you've written this comment but I did not realisee it. Thanks for catching it. Thanks Kaz for always catching such issues - they are hard to catch, and I appreciate how well you are checking the code, thanks!